client/virtual_channel_util.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2021 - See NOTICE file for copyright holders.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+	"math/big"
+	"sync"
+
+	"perun.network/go-perun/channel"
+)
+
+func (c *Channel) translateBalances(indexMap []channel.Index) channel.Balances {
+	state := c.state()
+	return transformBalances(state.Balances, state.NumParts(), indexMap)
+}
+
+func transformBalances(b channel.Balances, numParts int, indexMap []channel.Index) (_b channel.Balances) {
+	_b = make(channel.Balances, len(b))
+	for a := range _b {
+		_b[a] = make([]*big.Int, numParts)
+		// Init with zero.
+		for p := range _b[a] {
+			_b[a][p] = big.NewInt(0)
+		}
+		// Fill at specified indices.
+		for p, _p := range indexMap {
+			_b[a][_p] = b[a][p]
+		}
+	}
+	return
+}
+
+func (c *Client) rejectProposal(responder *UpdateResponder, reason string) {
+	ctx, cancel := context.WithTimeout(c.Ctx(), responseTimeout)
+	defer cancel()
+	err := responder.Reject(ctx, reason)
+	if err != nil {
+		c.log.Warnln(err)
+	}
+}
+
+func (c *Client) acceptProposal(responder *UpdateResponder) {
+	ctx, cancel := context.WithTimeout(c.Ctx(), responseTimeout)
+	defer cancel()
+	err := responder.Accept(ctx)
+	if err != nil {
+		c.log.Warnln(err)
+	}
+}
+
+type watcherEntry struct {
+	state interface{}
+	done  chan struct{}
+}
+
+type stateWatcher struct {
+	sync.RWMutex
+	entries   map[uint]watcherEntry
+	counter   uint
```

Comment:
I was considering that, but it has some downsides: https://forum.golangbridge.org/t/using-interface-types-as-map-keys/706

In particular, it would mean we have to be careful what we use the watcher for. Otherwise it might panic.

Reply:
In general, I'd agree that using interface map keys is a bad idea, e.g., if we were writing generic exported library code.

But here, we use the `state`s for a very specific user case, namely, for struct values, where it is totally fine. And since it would panic in the tests already if it wasn't fine, IMO it is ok to do it in this special use case. I mean, we don't even export the state watcher but only use it package-locally in very specific use cases.

Reply:
OK, created a corresponding commit and it seems to work.